## Data extracting and cleaning to map the counts of events to US states

In [2]:
# import relevant libraries
import pandas as pd
import numpy as np

In [12]:
# read data
data = pd.read_csv('../Data/ufo_clean.csv')

In [13]:
data.head()

Unnamed: 0  index year_week date_of_sight  year  month  week_of_year state  \
0           0      0   2006-52    2006-01-01  2006      1            52    NY   
1           1      1   2006-52    2006-01-01  2006      1            52    TX   
2           2      2    2006-1    2006-01-02  2006      1             1    MI   
3           3      3    2006-1    2006-01-02  2006      1             1    VA   
4           4      4    2006-1    2006-01-04  2006      1             1    PA   

                            city  \
0  I-80 (unknown city proximity)   
1   San Antonio/Laredo (between)   
2                      Ann Arbor   
3                 Fredericksburg   
4                       Scotland   

                                       sight_summary  ufo_shape  
0  Traveling from Brooklyn, NY to Groton, NY on I...  Formation  
1  Three separate sightins which ocurred in rural...   Triangle  
2  Probable HOAX: 5 ufos, silent black objects, m...   Changing  
3  clear sky looked like a star 10:00 am in the m...      Light  
4  Small, white, light in the northeast sky...dis...      Light

In [27]:
# only need a few columns for the map_analysis, therefore drop unnecessary columns
map_data = data.drop(columns=['Unnamed: 0', 'index', 'year_week', 'date_of_sight', 'month', 'week_of_year', 'city','sight_summary', 'ufo_shape'])
map_data
# it would have more efficient to pick only the relevant columns instead of dropping the unnecessary ones

year state
0      2006    NY
1      2006    TX
2      2006    MI
3      2006    VA
4      2006    PA
...     ...   ...
71560  2020    PA
71561  2020    PA
71562  2020    FL
71563  2020    MN
71564  2020    OR

[71565 rows x 2 columns]

In [40]:
# checking for NaN
map_data.isna().sum()

year       0
state    107
dtype: int64

In [41]:
# for 'year', there was no NaN, for 'state' there is 0,15% of the data missing 
# this is neglectable and therefore be dropped out of the dataset (107 rows are dropped)
map_data1 = map_data.dropna()
map_data1.isna().sum()

year     0
state    0
dtype: int64

In [43]:
# value counts
map_data1['state'].value_counts()

CA                            8494
FL                            4757
WA                            3501
TX                            3045
NY                            2992
                              ... 
 on Hwy S 29)                    1
PE                               1
 Muncies Marina)                 1
 Weld Co.)                       1
 near Wilderness Camp Rd.)       1
Name: state, Length: 189, dtype: int64

In [ ]:
# I want to see all the states
pd.set_option('display.max_rows', None)

In [50]:
# now I can see all entries in the state column
map_data1['state'].value_counts()

CA                                     8494
FL                                     4757
WA                                     3501
TX                                     3045
NY                                     2992
PA                                     2857
AZ                                     2546
OH                                     2426
IL                                     2234
NC                                     2188
MI                                     1977
OR                                     1908
CO                                     1759
NJ                                     1571
GA                                     1564
VA                                     1549
MO                                     1523
MA                                     1507
SC                                     1462
WI                                     1416
IN                                     1319
TN                                     1274
MN                              

There are 50 states in the US. There are 189 different states entries on the state column. 
Even with a few entries that repeated, e.g. 'in flight', 'on Interstate 5' etc., I will eliminate those entries.
In total, we might loose ca. 200 rows max. 200 out of over 71k rows accounts only for < 0,3% of data loss.

In [60]:
# make list of all US states
state_lst = ['AL', 'AK', 'AZ', 'AR', 'CA', 'CO', 'CT', 'DE', 'FL', 'GA', 'HI', 'ID', 'IL', 'IN', 'IA', 'KS', 'KY', 
             'LA', 'ME', 'MD', 'MA', 'MI', 'MN', 'MS', 'MO', 'MT', 'NE', 'NV', 'NH', 'NJ', 'NM', 'NY', 'NC', 'ND',
             'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX', 'UT', 'VT', 'VA', 'WA', 'WV', 'WI', 'WY']

# save only rows with a state entry which corresponds with one of the 50 states in the state list above
map_data2 = map_data1[map_data1['state'].apply(lambda x: x in state_lst)]
map_data2.shape

(71079, 2)

In [56]:
# save this dataframe to continue working in Tableau
map_data2.to_csv('../Data/map_data_tableau.csv')

In [58]:
# save value counts into a new dataframe
map_data3 = map_data2['state'].value_counts()
map_data3

CA    8494
FL    4757
WA    3501
TX    3045
NY    2992
PA    2857
AZ    2546
OH    2426
IL    2234
NC    2188
MI    1977
OR    1908
CO    1759
NJ    1571
GA    1564
VA    1549
MO    1523
MA    1507
SC    1462
WI    1416
IN    1319
TN    1274
MN    1200
CT    1135
MD    1071
KY     899
NV     897
NM     894
UT     877
ID     791
OK     781
AL     749
IA     697
NH     662
ME     656
KS     636
LA     598
AR     588
MT     561
WV     483
HI     407
VT     383
MS     378
RI     368
NE     359
AK     348
DE     242
SD     212
WY     199
ND     139
Name: state, dtype: int64

In [59]:
# save this dataframe to continue working in Tableau
map_data3.to_csv('../Data/map_data_tableau3.csv')